scripts/zkai_d64_external_adapter_surface_probe.py

```
@@ -0,0 +1,585 @@
+        "required_custom_semantics",
+        "next_action",
+    }
+    if set(candidate) != required:
+        raise D64ExternalAdapterProbeError("candidate adapter field set mismatch")
+    if candidate["proof_generated"] is not False:
+        raise D64ExternalAdapterProbeError("candidate adapters must not claim proof generation")
+    if not isinstance(candidate["required_custom_semantics"], list):
+        raise D64ExternalAdapterProbeError("candidate required_custom_semantics must be a list")
+
+
+def validate_probe(payload: dict[str, Any]) -> None:
```

Ruff (0.15.12)

[warning] Too many branches (22 > 12)

(PLR0912)